### PR TITLE
Model: make string attribute value auto-trimming optional

### PR DESCRIPTION
**Pull request by lex0r**
because not all attributes are made equal.

### CODE REVIEW

**Comment by LukeTowers** [comment]
@lex0r can you add the test case from the original PR? https://github.com/octobercms/library/pull/503/files#diff-270f442232291e973d322f9166dda6fb544ad2325f6db11581f7d553171213a8

Also I'm curious to know your use case and what problems auto trimming was causing for you, are you able to share some details?

File: src/Database/Model.php [hunk 1]
```diff
@@ -49,6 +49,11 @@ class Model extends EloquentModel
      */
     public $duplicateCache = true;
 
+    /**
+     * @var bool Indicates if all model attributes will be trimmed prior to saving.
+     */
+    public $trimAttributes = true;
```
Comment (LukeTowers) [inline src/Database/Model.php]:
```suggestion
    /**
     * @var bool Indicates if all string model attributes will be trimmed prior to saving.
     */
    public $trimStringAttributes = true;
```

File: src/Database/Model.php [hunk 2]
```diff
@@ -1239,7 +1244,7 @@ public function setAttribute($key, $value)
         /*
          * Trim strings
          */
-        if (is_string($value)) {
+        if ($this->trimAttributes && is_string($value)) {
```
Comment (LukeTowers) [inline src/Database/Model.php]:
```suggestion
        if ($this->trimStringAttributes && is_string($value)) {
```

**Comment by RomainMazB** [comment]
> @lex0r can you add the test case from the original PR? https://github.com/octobercms/library/pull/503/files#diff-270f442232291e973d322f9166dda6fb544ad2325f6db11581f7d553171213a8
> 
> Also I'm curious to know your use case and what problems auto trimming was causing for you, are you able to share some details?

Great addition here!!

As I remember, my use case was forum messages including markdown: if you start a message with a code block using tab instead of three backticks, the tabs was trimmed and the code block just didn't work...

**Comment by lex0r** [comment]
> @lex0r can you add the test case from the original PR? https://github.com/octobercms/library/pull/503/files#diff-270f442232291e973d322f9166dda6fb544ad2325f6db11581f7d553171213a8
> 
> Also I'm curious to know your use case and what problems auto trimming was causing for you, are you able to share some details?

@LukeTowers I had to save a file to a remote location directly without any pre-processing and then save a record linking to it in the database. The file name can contain trailing spaces and that is acceptable, however when trying to save such a name in the database you would get the name trimmed. Later on we used the name to construct a link to the file and potential spaces were not included in that link which made it impossible to locate the file in the remote storage which has the file saved with the space.

**Comment by LukeTowers** [comment]
Interesting, thanks for sharing your use case @lex0r!

**Pull request merged**